src/coreclr/vm/virtualcallstub.cpp

```diff
@@ -1114,27 +1268,61 @@ BYTE *VirtualCallStubManager::GenerateStubIndirection(PCODE target, BOOL fUseRec
     // Allocate from loader heap
     if (!ret)
     {
+        size_t alignment;
+        INTERFACE_DISPATCH_CACHED_OR_VSD(alignment = sizeof(TADDR) * 2, alignment = sizeof(TADDR));
+
         // Free list is empty, allocate a block of indcells from indcell_heap and insert it into the free list.
-        BYTE ** pBlock = (BYTE **) (void *) indcell_heap->AllocMem(S_SIZE_T(cellsPerBlock) * S_SIZE_T(sizeof(BYTE *)));
+        size_t cellsAllocationSize = cellsPerBlock * sizeOfIndCell;
+        size_t allocationSize = cellsAllocationSize;
+
+#ifdef FEATURE_CACHED_INTERFACE_DISPATCH
+        if (m_loaderAllocator->IsCollectible() && UseCachedInterfaceDispatch())
+        {
+            allocationSize += sizeof(CachedIndirectionCellBlockListNode);
+        }
+#endif // FEATURE_CACHED_INTERFACE_DISPATCH
+        BYTE ** pBlock = (BYTE **) (void *) indcell_heap->AllocAlignedMem(cellsPerBlock * sizeOfIndCell, alignment);
+
+#ifdef FEATURE_CACHED_INTERFACE_DISPATCH
+        if (m_loaderAllocator->IsCollectible() && UseCachedInterfaceDispatch())
+        {
+            CachedIndirectionCellBlockListNode * pBlockNode = (CachedIndirectionCellBlockListNode *)pBlock;
+            pBlockNode->m_pNext = m_indirectionBlocks;
+            m_indirectionBlocks = pBlockNode;
+            pBlock = (BYTE **)(&pBlockNode->m_rgIndCells[0]);
+        }
+#endif // FEATURE_CACHED_INTERFACE_DISPATCH
 
         // return the first cell in the block and add the rest to the free list
         ret = (BYTE *)pBlock;
 
         // link all the cells together
         // we don't need to null terminate the linked list, InsertIntoFreeIndCellList will do it.
+        BYTE** pBlockCur = pBlock;
         for (UINT32 i = 1; i < cellsPerBlock - 1; ++i)
         {
-            pBlock[i] = (BYTE *)&(pBlock[i+1]);
+            BYTE** pBlockNext = (BYTE**)(((BYTE*)pBlockCur) + sizeOfIndCell);
+            *pBlockCur = (BYTE *)pBlockNext;
+            pBlockCur = (BYTE**)pBlockNext;
         }
 
         // insert the list into the free indcell list.
-        InsertIntoFreeIndCellList((BYTE *)&pBlock[1], (BYTE*)&pBlock[cellsPerBlock - 1]);
+        InsertIntoFreeIndCellList((((BYTE*)pBlock) + sizeOfIndCell), (((BYTE*)pBlock) + ((cellsPerBlock - 1) * sizeOfIndCell)));
```

Comment:
I'd love to see this simplified/abstracted a bit so it's clearer what's going on with all the pointer arithmetic and conversions